mne/time_frequency/csd.py

```diff
@@ -1346,5 +1347,51 @@ def _csd_morlet(data, sfreq, wavelets, nfft, tslice=None, use_fft=True,
 
     # Scaling by sampling frequency for compatibility with Matlab
     csds /= sfreq
-
     return csds
+
+
+@verbose
+def compute_csd(epochs_tfr, verbose=None):
+    """Compute covariance matrices across frequencies for TFR epochs.
+
+    Parameters
+    ----------
+    epochs_tfr : EpochsTFR
+        The time-frequency resolved epochs over which to compute the
+        covariance.
+    %(verbose)s
+
+    Returns
+    -------
+    res : instance of CrossSpectralDensity
+        Cross-spectral density restricted to selected channels.
+    """
+    _validate_type(epochs_tfr, EpochsTFR)
+    n_channels, n_freqs = len(epochs_tfr.ch_names), epochs_tfr.freqs.size
+    data = np.zeros((n_channels * (n_channels + 1) // 2, n_freqs),
+                    dtype=np.complex128)
+    # iterate over epochs
+    for idx, epochs_data in enumerate(epochs_tfr.data):
+        epochs_data_conj = np.conj(epochs_data)
+        # This is equivalent to:
+        # csds = np.vstack([np.mean(epochs_data[[i]] * epochs_data_conj[i:],
+        #                           axis=2) for i in range(n_channels)])
+        # There is a redundancy in the calculation here because we don't really
+        # need the lower triangle of the matrix, but it should still be faster
+        # than a loop (hopefully!).
+        csds = np.einsum('xft,yft->xyf', epochs_data, epochs_data_conj)
+        csds = csds[np.triu_indices(len(epochs_data)) + (slice(None),)]
+        csds /= epochs_data.shape[-1]
+
+        # Scaling by sampling frequency for compatibility with Matlab
+        csds /= epochs_tfr.info['sfreq']
+        data += csds
+
+    # scale to compute mean
+    data /= len(epochs_tfr)
+
+    # TO DO: EpochTFR should store n_fft and projs to be consistent
```

Comment:
Does `EpochsTFR` really not have a `.info` attribute? If it has it, then projs should be in `epochs_tfr.info['projs']` already

Reply:
It does have an info, I thought when I checked there wasn't a projs key in info which I thought was odd... I'll double check

Reply:
You might need to do `info.get('projs', None)` because I think sometimes the key might just not exist. But if you compute on an instance like one from `sample` that *does* have projs and this ends up as `None` then it's a problem...

Reply:
That's probably what happened, it was I think made from scratch so that's probably why it didn't even have a `projs` field.